### PR TITLE
Stack-safe FreeAp

### LINKED ISSUE #1151: Make FreeApplicative stack safe

### PR DESCRIPTION
**Pull request by edmundnoble**
Port from JS, from @safareli's implementation. It's stack-safe, I just need to benchmark and comment it.

Fixes #1151.

### CODE REVIEW

File: docs/src/main/tut/datatypes/freeapplicative.md [hunk 1]
```diff
@@ -164,5 +164,24 @@ val prodCompiler: FunctionK[ValidationOp, ValidateAndLog] = parCompiler and logC
 val prodValidation = prog.foldMap[ValidateAndLog](prodCompiler)
 ```
 
+### The way FreeApplicative#foldMap works
+Despite being an imperative loop, there is a functional intuition behind `FreeApplicative#foldMap`.
+
+The new `FreeAp`'s `foldMap` is a sort of mutually-recursive function that operates on an argument stack and a
```
Comment (kailuowang) [inline docs/src/main/tut/datatypes/freeapplicative.md]: quick nitpick question: shall we use `FreeAp` or `FreeApplicative` in the documentation?

Comment (edmundnoble) [inline docs/src/main/tut/datatypes/freeapplicative.md]: Good point, I'll try to use `FreeApplicative`. My bad.

File: free/src/main/scala/cats/free/FreeApplicative.scala [hunk 2]
```diff
@@ -4,82 +4,196 @@ package free
...
+  }
+  // scalastyle:on method.length
+
+
   /** Interpret/run the operations using the semantics of `Applicative[F]`.
-   * Tail recursive only if `F` provides tail recursive interpretation.
-   */
-  final def fold(implicit F: Applicative[F]): F[A] =
+    * Tail recursive only if `F` provides tail recursive interpretation.
+    */
```
Comment (kailuowang) [inline free/src/main/scala/cats/free/FreeApplicative.scala]: another nitpick here, the scala doc indentation is changed away from our standard.

File: free/src/main/scala/cats/free/FreeApplicative.scala [hunk 3]
```diff
@@ -4,82 +4,196 @@ package free
...
+  final def fold(implicit F: Applicative[F]): F[A] = {
     foldMap(FunctionK.id[F])
+  }
```
Comment (kailuowang) [inline free/src/main/scala/cats/free/FreeApplicative.scala]: yet another superficial comment, what about all these addition of `{`. If we want to change the formating rule, shall we use another separate PR?

File: free/src/test/scala/cats/free/FreeApplicativeTests.scala [hunk 4]
```diff
@@ -56,6 +64,17 @@ class FreeApplicativeTests extends CatsSuite {
     r1.foldMap(nt) should === (r2.foldMap(nt))
   }
 
+  test("FreeApplicative#flatCompile") {
```
Comment (kailuowang) [inline free/src/test/scala/cats/free/FreeApplicativeTests.scala]: How about making it a scalacheck?

**Comment by kailuowang** [comment]
I left a couple of superficial comments. To me, it's a bit too magical to validate eyeball, I think making all relevant tests more robust (by converting them to scalacheck) would help. I am also curious about the benchmarks.

File: free/src/main/scala/cats/free/FreeApplicative.scala (hunk 2 above)
Comment (johnynek) [inline free/src/main/scala/cats/free/FreeApplicative.scala]: I don't think I know what tail recursive implementation means for Applicative. Can we be more concrete or link somewhere in this comment?

Comment (edmundnoble) [inline free/src/main/scala/cats/free/FreeApplicative.scala]: It's stack-safe.

File: free/src/main/scala/cats/free/FreeApplicative.scala (hunk 3 above)
Comment (edmundnoble) [inline free/src/main/scala/cats/free/FreeApplicative.scala]: IntelliJ is conspiring against me 😛

File: free/src/test/scala/cats/free/FreeApplicativeTests.scala (hunk 4 above)
Comment (edmundnoble) [inline free/src/test/scala/cats/free/FreeApplicativeTests.scala]: I'll do that, and add an Arbitrary which is more exhaustive.

Comment (edmundnoble) [inline free/src/test/scala/cats/free/FreeApplicativeTests.scala]: Doing this has exposed a `ClassCastException` bug in the implementation, which I'm working through now.

File: free/src/main/scala/cats/free/FreeApplicative.scala (hunk 3 above)
Comment (edmundnoble) [inline free/src/main/scala/cats/free/FreeApplicative.scala]: Fixed.

File: free/src/main/scala/cats/free/FreeApplicative.scala (hunk 2 above)
Comment (edmundnoble) [inline free/src/main/scala/cats/free/FreeApplicative.scala]: Done.

File: free/src/test/scala/cats/free/FreeApplicativeTests.scala (hunk 4 above)
Comment (edmundnoble) [inline free/src/test/scala/cats/free/FreeApplicativeTests.scala]: Fixed it.

**Comment by edmundnoble** [comment]
@kailuowang I did some basic benchmarks. A right-associated call tree with 50000 `Ap` calls takes as long to construct and evaluate now as one with 1000 leaves did previously. We're at 50x faster, presumably mostly because `Ap` used to take O(n) time and now takes O(1).

I also converted the tests I thought were appropriate to scalacheck.

**Comment by safareli** [comment]
> @kailuowang I did some basic benchmarks. A right-associated call tree with 50000 Ap calls takes as long to construct and evaluate now as one with 1000 leaves did previously. We're at 50x faster, presumably mostly because Ap used to take O(n) time and now takes O(1). I tried evaluating the same 50000 Ap tree using the old implementation, and my system ran out of RAM 😛



Perfect 🎉  💃

File: docs/src/main/tut/datatypes/freeapplicative.md (hunk 1 above)
Comment (kailuowang) [inline docs/src/main/tut/datatypes/freeapplicative.md]: looks like you forgot about this one?

File: laws/src/main/scala/cats/laws/discipline/Arbitrary.scala [hunk 5]
```diff
@@ -165,6 +166,22 @@ object arbitrary extends ArbitraryInstances0 {
 
   implicit def catsLawsArbitraryForReaderWriterStateT[F[_]: Applicative, E, L, S, A](implicit F: Arbitrary[(E, S) => F[(L, S, A)]]): Arbitrary[ReaderWriterStateT[F, E, L, S, A]] =
     Arbitrary(F.arbitrary.map(ReaderWriterStateT(_)))
+
+  implicit def catsLawsArbitraryForListNatTrans: Arbitrary[List ~> List] =
```
Comment (kailuowang) [inline laws/src/main/scala/cats/laws/discipline/Arbitrary.scala]: These two arbitrary instances seems a bit too, well, arbitrary and limited to be included in discipline for general uses. (actually the second one might be unused according to the codcov), how about we move them into the tests?

Comment (edmundnoble) [inline laws/src/main/scala/cats/laws/discipline/Arbitrary.scala]: I'll put the used one into `FreeApplicativeTests`.

Comment (edmundnoble) [inline laws/src/main/scala/cats/laws/discipline/Arbitrary.scala]: Done.

**Comment by kailuowang** [comment]
Thanks. The only minor comment left from my side is the `FreeAp` wording in the documentation.

**Comment by edmundnoble** [comment]
Oh wow, thought I fixed that a while ago. Done.

File: free/src/test/scala/cats/free/FreeApplicativeTests.scala [hunk 6]
```diff
@@ -126,3 +129,44 @@ class FreeApplicativeTests extends CatsSuite {
...
+      f <- freeGen[F, A](fDepth).map(_.map(l => (u: A) => ff(l, u)))
+      freeFA <- freeGen[F, A](freeDepth)
+    } yield freeFA.ap(f)
+
+    if (maxDepth <= 1) noFlatMapped
+    else Gen.oneOf(noFlatMapped, withFlatMapped)
+  }
+
+  implicit def freeArbitrary[F[_], A](implicit F: Arbitrary[F[A]], FF: Arbitrary[(A, A) => A], A: Arbitrary[A]): Arbitrary[FreeApplicative[F, A]] =
+    Arbitrary(freeGen[F, A](2))
```
Comment (kailuowang) [inline free/src/test/scala/cats/free/FreeApplicativeTests.scala]: Sorry I meant to ask this last time but somehow lost it. Did you try `maxDepth` larger than 2? The main reason I ask here is that it seemed to me that you were preparing for testing maxDepth more than 2.

**Comment by kailuowang** [comment]
@peterneyens @johnynek since you two are the only other maintains left around, got some time to review this one? It would nice to be included in the MF release.

File: free/src/test/scala/cats/free/FreeApplicativeTests.scala (hunk 6 above)
Comment (edmundnoble) [inline free/src/test/scala/cats/free/FreeApplicativeTests.scala]: Yeah I was, I forgot to raise it again after lowering.

Comment (edmundnoble) [inline free/src/test/scala/cats/free/FreeApplicativeTests.scala]: Fixed.

**Comment by peterneyens** [comment]
This indeed feels a bit magical.

I played around a bit to get a feel what is going on and end up refactoring `foldMap` a bit. I haven't done any benchmarks to compare it, but it seems to pass the tests.
It is probably a bit slower, but I think it might be easier to comprehend than the more imperative implementation.

See https://github.com/edmundnoble/cats/compare/stacksafefreeap...peterneyens:stacksafefreeap-peter

**Comment by edmundnoble** [comment]
@peterneyens Wow, thank you, this is really readable. It looks like it has to be noticeably less performant than the imperative implementation (extra intermediate list which is appended to the right and extra tuple allocation introduced while reassociating an `Ap` chain) but I think I can use this to improve the comments a lot.

**Comment by peterneyens** [comment]
Thanks for the improved documentation @edmundnoble.

Given the considerable performance improvements and the added stack safety, lets merge this as is, so we don't have to hold the 1.0.0-MF release any longer :+1: (Looks like we need to fix some conflicts first though).

We can look if we can clean up the `foldMap` implementation later on.

**Comment by kailuowang** [comment]
merging with 2 sign-offs

**Pull request merged**